pkgs/development/libraries/ffmpeg/generic.nix

```
@@ -4,10 +4,10 @@
   # NOTICE: Always use this argument to override the version. Do not use overrideAttrs.
 , version # ffmpeg ABI version. Also declare this if you're overriding the source.
 , hash ? "" # hash of the upstream source for the given ABI version
+, rev ? "" # git revision of the upsteam source
 , source ? fetchgit {
     url = "https://git.ffmpeg.org/ffmpeg.git";
-    rev = "n${version}";
-    inherit hash;
+    inherit rev hash;
```

This must be kept as it's part of the external interface.

I'd just do `rev ? null` and then

```suggestion
    rev = if rev != null then rev else "n${version}";
    inherit hash;
```

---

You're right! But should it be allowed to use the external interface in a way that guarantees that the build is broken or insecure? (Not providig `rev` leads to a broken/insecure `$n{version}$` build.)

The only instance in Nixpkgs where the interface is called without providing `rev` is in `ffmpeg/default.nix`.

If some future package uses ffmpeg without providing `rev`, maybe an easy-to-fix build failure is preferred over a broken/insecure build?

---

> should it be allowed to use the external interface in a way that guarantees that the build is broken or insecure?

That's up to the user to decide; this merely implements what they ask it to do.

> The only instance in Nixpkgs where the interface is called without providing `rev` is in `ffmpeg/default.nix`.

I know of at least one instance which is handbrake's ffmpeg. Their usage of this interface is a reason I'd almost slightly prefer applying patches but ugh.

> If some future package uses ffmpeg without providing `rev`, maybe an easy-to-fix build failure is preferred over a broken/insecure build?

No, it's squarely their responsibility to ensure the version they chose (rather than the default we provide) is secure for their purposes IMHO.

---

First and foremost, I want to thank you for your time and effort to explain all of this! I am trying to understand how package maintainers are handling vulnerabilities, so I really appreciate your insights.

I agree with your points, but I do want to highlight a few things:
* The interface would still implement what the user asks it to do, as long as `rev` is provided explicitly. The functionality of `ffmpeg/generic.nix` is not restricted. (And of course overriding the `ffmpeg` source as done in ffmpeg-livepeer, jellyfin-ffmpeg, and other packages is also still possible and builds without any problems.)
* You're right that HandBrake is not providing `rev`, but your proposed change would not set it to `"n${version}"` either. (We have `rev == v7.rev` as its value is already set by `ffmpeg_7-full`. So what happens for HandBrake is that `if rev != null then rev else "n${version}"` evaluates to `rev=rev`, which is also what `inherit rev` would do.)
* If you really want `ffmpeg/generic.nix` to support insecure/broken `n${version}` builds, then your proposed change would not be sufficient. The n7.1 build fails without `./fix-fate-ffmpeg-spec-disposition-7.1.patch`, which I removed because `release/7.1` no longer needs it.

Having said this, I think we all agree that there are no good solutions here, just different levels of headache because ffmpeg upstream is apparently fixing serious bugs without bumping version numbers...